### ISSUE TITLE
Can't work with optional array value in conditional type condition

### ISSUE THREAD

**Issue by danog** [vimeo/psalm#6891]
See https://psalm.dev/r/9c71795861
Psalm doesn't properly interpret the `TReturnOpt is empty` expression, leading to weird bugs in return types.
The type of `TReturnOpt` is indeed `empty' (TEmpty) if the array value is not present, as verified in this example: https://psalm.dev/r/40fdaaba10

Here's a real-life-inspired example: https://psalm.dev/r/b1240bf321

**Comment by danog** [comment]
Using `null` instead of `empty` in the comparison is a workaround: https://psalm.dev/r/9fd6adea45
However, there is still some strange non-documented behavior that leads to the selection of the first (?) type of a conditional chain if the type is `empty`, as shown in the first examples.

**Comment by orklah** [comment]
I believe you need `array<empty, empty>`: https://psalm.dev/r/0d26ebbe04

**Comment by danog** [comment]
Actually, no, `array<empty, empty>` should not work at all, since `ReturnOpt`  is extracted from the array element `ret`, with the only allowed values being "`string"|"int"|"float"` (no `array` types!)

**Comment by orklah** [comment]
Right, I wanted to post something like https://psalm.dev/r/5e238c29a1 instead, but there's something wrong too.

**Comment by orklah** [comment]
Well, the first report is not really a bug, empty is not made to be used for that.

I'd actually suggest using `never` instead, but it doesn't work either because `never` is the bottom type so it's indeed contained in `string` or `int` or any other type (it's a subtype of every type)

My answer as you pointed is indeed weirder and the last link I posted is weird too.

**Comment by orklah** [comment]
For future reference, here are simple reproducers:
this should work: https://psalm.dev/r/50024e6265
this shouldn't: https://psalm.dev/r/bedb1f12b2

**Comment by orklah** [comment]
My first reproducer is wrong, when you take the code like that, Psalm translate that to `$a['ret'] = isset($a['ret']) ? $a['ret'] : 'a';`

And then you see that it is indeed a MixedAssignment because Psalm doesn't know what's in there.

But it's also normal that in the original, Psalm emits a MixedAssignment because $a['ret'] is possibly undefined, which means Psalm has no idea what it will get back, so no bugs here.

I'll look at the second one
